pkg/netns/netns_linux.go

```diff
@@ -0,0 +1,200 @@
+// Note that FDs should not be used for comparing or testing equality, for that
+// see `equals`
+func (h *NetNSHandle) FD() int {
+	return h.fd
+}
+
+// Close closes the handle to the network namespace. Note that this does not guarantee
+// the network namespace itself will be destroyed or closed, which will happen when all references to it
+// (including the one being closed here) are gone.
+func (h *NetNSHandle) Close() error {
+	if err := unix.Close(h.fd); err != nil {
+		return err
+	}
+	h.fd = -1
+	return nil
+}
+
+// Do runs the provided func within a new thread+netns, without changing the current thread's netns
+//
+// Note that since network namespaces are set *per thread*, the thread the closure is running inside
+// will be locked with `runtime.LockOSThread()` during this operation.
+//
+// This means the provided closure should _not_ call runtime.UnlockOSThread() at all,
+// or else undefined behavior should be expected.
+func (h *NetNSHandle) Do(runInside func(*NetNSHandle) error) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// Start the callback in a new green thread so that if we later fail
+	// to switch the namespace back to the original one, we can safely
+	// leave the thread locked to die without a risk of the current thread
+	// left lingering with incorrect namespace.
+	var innerError error
+	go func() {
+		// save a handle to current network namespace
+		// and lock the thread - this will force this goroutine to be
+		// scheduled in a different thread, so we should be safe to mutate the netns
+		restore, err := saveRestoreCurrentThreadNetns()
+		if err != nil {
+			innerError = err
+			return
+		}
+		defer restore()
+		defer wg.Done()
+		unix.Setns(h.fd, unix.CLONE_NEWNET)
+		innerError = runInside(h)
+	}()
+	wg.Wait()
+
+	return innerError
+}
+
+// NOTE If we ever need to test equality of netNSes outside of test code,
+// make this public.
+func (h *NetNSHandle) equal(other *NetNSHandle) bool {
+	if h.FD() == other.FD() {
+		return true
+	}
+
+	// Just because the FDs are not equal doesn't mean they don't both refer
+	// to the same underlying netns, so check inode too.
+	var s1, s2 unix.Stat_t
+	if err := unix.Fstat(int(h.FD()), &s1); err != nil {
+		return false
+	}
+	if err := unix.Fstat(int(other.FD()), &s2); err != nil {
+		return false
+	}
+	return (s1.Dev == s2.Dev) && (s1.Ino == s2.Ino)
+}
+
+// saveRestoreCurrentThreadNetns returns a cleanup func (which callers should `defer`)
+// and an error if for some reason the current netns is unobtainable.
+//
+// the cleanup func will move the thread back to the netns it was running under when
+// this function was called, and unlock the current OS thread.
+func saveRestoreCurrentThreadNetns() (func(), error) {
+	runtime.LockOSThread()
+	hostNSFD, err := getCurrent()
+	if err != nil {
+		// flaking out early, so unlock current thread
+		runtime.UnlockOSThread()
+		return nil, fmt.Errorf("Failed to open current namespace: %v", err)
```

Comment:
Nit: errors imply failure. Also, avoid caps at the start of error strings, they're hard to parse when the error gets wrapped. Also, `%w` is the verb for wrapping errors.
```suggestion
		return nil, fmt.Errorf("get current namespace: %w", err)
```